keps/sig-storage/3294-provision-volumes-from-cross-namespace-snapshots/README.md

```diff
@@ -0,0 +1,1024 @@
+
+There will be two approaches to implement as a separate populator:
+- [(1) Populate data from snapshot to provisioned PV](#1-populate-data-from-snapshot-to-provisioned-pv)
+- [(2) Provision PV with data via CSI call](#2-provision-pv-with-data-via-csi-call)
+
+##### (1) Populate data from snapshot to provisioned PV
+
+This is a straightforward implementation that AnyVolumeDataSource feature defines.
+Developers will be able to utilize lib-volume-populator to implement this way.
+One of the challenges to achieve it will be how to actually copy the data from a snapshot in one namespace to an already provisioned PV that will need to be bound to a PVC in the other namespace.
+
+A naive implementation will be:
+1. Create another PV from the snapshot in the snapshot's namespace,
+2. Copy the data from the PV to somewhere accessible from any namespaces,
+3. Copy the data in step 2 to the originally intended PV,
+4. Delete the temporary data in step 1 and step 2.
+
+If the naive implementation is used, unintended transient states, for example a temporary PVC in the snapshot namespace, may be visible to users.
+Also, there may be performance issues depending on where and how data is copied.
+
+On the other hand, althoguh it completely depends on the implementation, this approach can have advantages, like the ability to populate volumes from snapshot across different CSI drivers or the ability to efficiently copy data by using CSI driver specific way.
+
+There will be no generic way to implement by using this approach, because the implementations rely too much on backup tools or CSI drivers.
+Therefore no community implementation of this approach will be provided.
+
+##### (2) Provision PV with data via CSI call
+
+Current CSI external provisioners provision volume regardless of the data source, therefore populators need to populate data to already provisioned PVs.
+However, this behavior may be changed and provisioners may offload provisioning to populators for PV with `VolumeSnapshotLink` CRD data source.
+
+The implementation of provisioner and populator of this approach will be as follows:
+
+- Provisioner:
+  - Handles `VolumeSnapshotLink` CRD,
+  - Checks if `VolumeSnapshotLink` is specified as `DataSourceRef`:
+    - If specified, skip provisioning  the volume
+
+- Populator:
+  - Handles `VolumeSnapshotLink` CRD and `ReferencePolicy` CRD,
+  - Checks if `VolumeSnapshotLink` is specified as `DataSourceRef`:
+    - If specified, check if the access to the `VolumeSnapshot` referenced by the `VolumeSnapshotLink` is allowed by any `ReferencePolicy`s:
+      - If allowed, use the `VolumeSnapshot` as a SnapshotSource to pass to the CSI driver for provision.
+
+The above implementation is just separating the logics in approach (a) to two components, and it won't help improve efficiency nor simplify implementations.
+Therefore, the description in this section is just for discussion purpose and won't be implemented.
+
+### Test Plan
```

Comment:
As a small reminder, we have [updated](https://github.com/kubernetes/enhancements/pull/3279) our [KEP template Test Plan section](https://github.com/kubernetes/enhancements/tree/master/keps/NNNN-kep-template#test-plan) this enhancement is using the previous version. Please be sure to update relevant sections for acceptance into the release. :)

Reply:
@kikisdeliveryservice Updated the Test plan section with the new KEP template. Thank you for your notification on the template update.